### PR TITLE
fix: Resend stream data on `ZeroRttRejected`

### PR DESCRIPTION
**Pull request by larseggert**
Make the neqo client re-initiate 0-RTT requests that were rejected by the server.

### CODE REVIEW

**Comment by larseggert** [comment]
This seems to be due to this line
https://github.com/mozilla/neqo/blob/e3d8be487989f8fd53a04cc25162bffff63c91ba/neqo-transport/src/streams.rs#L103-L104
which causes all streams to be blown away on the client when 0-RTT is rejected by the server.

Removing this line causes two `zero_rtt_send_reject` tests to fail, so this looks like blowing the streams (and their data) away is expected behavior? I'm kinda at a loss here why this is the case; any insights @martinthomson?

**Comment by mxinden** [comment]
As far as I can tell, this is required by RFC 9001 @larseggert:

> When 0-RTT is rejected, all connection characteristics that the client assumed might be incorrect. This includes the choice of application protocol, transport parameters, and any application configuration. The client therefore MUST reset the state of all streams, including application state bound to those streams.

https://www.rfc-editor.org/rfc/rfc9001.html#section-4.6.2

**Comment by larseggert** [comment]
Right. But at least I understood "reset the state of all streams" to mean "reset what you think has been sent/acked on each stream" and not "make it so the streams never existed".

**If** what we do is correct, we need the client to correctly handle this, i.e., manually RTX that data.

File: neqo-bin/src/client/http3.rs [hunk 1]
```diff
@@ -222,7 +222,9 @@ impl<'a> super::Handler for Handler<'a> {
                     }
                 }
                 Http3ClientEvent::StateChange(Http3State::Connected)
-                | Http3ClientEvent::RequestsCreatable => {
+                | Http3ClientEvent::RequestsCreatable
+                | Http3ClientEvent::ZeroRttRejected => {
+                    qinfo!("{event:?}");
```
Comment (mxinden) [inline neqo-bin/src/client/http3.rs:227]: :+1: good idea!

Comment (larseggert) [reply neqo-bin/src/client/http3.rs:227]: I think for the interop runner, only h09 matters.

We should also check that Fx is doing the right thing.

Comment (KershawChang) [reply neqo-bin/src/client/http3.rs:227]: In Fx, we rewind the stream to the state it was in before attempting 0RTT. The code that handles the ZeroRttRejected event is [here](https://searchfox.org/mozilla-central/rev/527d691a542ccc0f333e36689bd665cb000360b2/netwerk/protocol/http/Http3Session.cpp#608). Essentially, we reset the [offset of the request stream](https://searchfox.org/mozilla-central/rev/527d691a542ccc0f333e36689bd665cb000360b2/netwerk/protocol/http/nsHttpTransaction.cpp#3079) and [all internal states of HTTP3Stream](https://searchfox.org/mozilla-central/rev/527d691a542ccc0f333e36689bd665cb000360b2/netwerk/protocol/http/Http3Stream.cpp#465-479).

I think this behavior is correct.

Comment (mxinden) [reply neqo-bin/src/client/http3.rs:227]: > I think for the interop runner, only h09 matters.

Same on h09, here I would expect the client to eventually call `self.download_urls(client)` given the match on:

``` rust
 ConnectionEvent::StateChange(
    State::WaitInitial | State::Handshaking | State::Connected,
 ) => {
```

**Review by martinthomson — COMMENTED** [review]
Keep in mind that a rejection of 0-RTT could be due to a reduction in the number of streams that you are allowed to create.  You can't just rewind, you have to start over completely.

File: neqo-bin/src/client/http09.rs [hunk 2]
```diff
@@ -76,7 +76,8 @@ impl<'a> super::Handler for Handler<'a> {
                 }
                 ConnectionEvent::StateChange(
                     State::WaitInitial | State::Handshaking | State::Connected,
-                ) => {
+                )
+                | ConnectionEvent::ZeroRttRejected => {
                     qdebug!("{event:?}");
                     self.download_urls(client);
```
Comment (martinthomson) [inline neqo-bin/src/client/http09.rs:82]: I don't think that this is correct.  You will need to reset the state of each of the streams as well.

My understanding is that if we have sent a request, we track that.  So if a request is partially sent, we need to resend.

(Maybe h09 doesn't do that, but h3 might have to.)

File: neqo-bin/src/client/http3.rs (hunk 1 above)
Comment (larseggert) [reply neqo-bin/src/client/http3.rs]: I think the issue (for both h3 and h09) is that already-tried URLs are popped off the list and are not retried. I have a fix for this now in the PR; CI is running.

File: neqo-bin/src/client/http09.rs [hunk 3]
```diff
@@ -80,6 +89,11 @@ impl<'a> super::Handler for Handler<'a> {
...
                 }
+                ConnectionEvent::ZeroRttRejected => {
```
Comment (mxinden) [inline neqo-bin/src/client/http09.rs:92]: Maybe add a comment here why all pending requests need to be retried?

File: neqo-bin/src/client/mod.rs [hunk 4]
```diff
@@ -351,6 +351,7 @@ trait Handler {
 
     fn handle(&mut self, client: &mut Self::Client) -> Res<bool>;
     fn take_token(&mut self) -> Option<ResumptionToken>;
+    fn reinit(&mut self);
```
Comment (mxinden) [inline neqo-bin/src/client/mod.rs:354]: `reinit` is always called on the concrete h09 or h3 Handler, never on the `Handler` trait, right? If so, I suggest removing the trait extension here and instead implement `reinit` on the concrete Handler types (e.g. `impl <'a> Handler<'a> { fn reinit(&mut self) {`) instead.

Not a strong opinion.

**Review by mxinden — APPROVED** [review]
Change looks good to me. Thank you for the debugging work!

**Review by mxinden — APPROVED** [review]

**Pull request merged**